### ISSUE TITLE
Intermittent deadlock due to fix for #719

### ISSUE THREAD

**Issue by rdasgupt** [eclipse-paho/paho.mqtt.java#763]
Please fill out the form below before submitting, thank you!

- [ ] Bug exists Release Version 1.2.2 ( Master Branch)
- [X] Bug exists in MQTTv3 Client on Snapshot Version 1.2.3-SNAPSHOT (Develop Branch)
- [ ] Bug exists in MQTTv5 Client on Snapshot Version 1.2.3-SNAPSHOT (Develop Branch)

If this is a bug regarding the Android Service, please raise the bug here instead: https://github.com/eclipse/paho.mqtt.android/issues/new

Opening this issue based on the comments from @marcostorto in issue #719  

--------

We found a nasty side effect of this fix. In our environment we use clean_session=false and the broker (vernemq) can take quite some time to respond to connect message when there are many (100K to 1M) messages waiting to be delivered.
In this case the client never completes the connection as MqttException.REASON_CODE_CONNECTION_LOST is raised before the broker respond to connect message. Retrying connection does not help, we get in a sort of deadlock as subsequent connection attempts fail in the same way.

I would strongly advice to remove this fix until this side effect is resolved, I assume it can happen with other brokers as well

We use alpakka library which uses async client [https://github.com/akka/alpakka/blob/master/mqtt/src/main/scala/akka/stream/alpakka/mqtt/impl/MqttFlowStage.scala#L147]

We can reproduce pretty easily by the following sequence:

    connect to vernemq with clean session=false
    subscribe to a topic with QOS=1
    terminate client
    publish few 100Ks messages to this topic (actual number of message to show isssue is somehow dependent on message size, for small message size issue appear >1M messages)
    reconnect client to broker using same client id

removing this commit connection completes successfully, with this commit connection fails most of the times (although not always).

**Comment by rdasgupt** [comment]
@ioolkos I have copied your comment from #719 

@marcostorto let me know when you want to look into this VerneMQ side. Most helpful in this case is a Wireshark capture between the client and broker (if that's possible).
One thing to try server side: we have seen that increasing the message inflight window in situations like this might help (form the default 20 to something much higher)

**Comment by rdasgupt** [comment]
@ioolkos @marcostorto
There is a new service release v1.2.3. Please try your use case again and let us know if this problem still exist.

**Comment by marcostorto** [comment]
I run an experiment using new release v1.2.3 in our test setup:

1. start consumer application (which connects 23 different clients to mqtt broker subscribing to multiple queues), let it subscribe with qos=1 and clean_session=false, then stop it
2. start producer applications
3. let about 1.3M messages accumulate in vernemq queues (3 queues with about 400K messages each, plus another queue with about 50K messages)
4. restart the consumer application 

after restart most clients connect without issues, but clients which have many messages in queue disconnect multiple times from broker.
I attach here a tcpdump capture for a failed connection [bad.zip](https://github.com/eclipse/paho.mqtt.java/files/4482458/bad.zip) and for a successfull connection [good.zip](https://github.com/eclipse/paho.mqtt.java/files/4482457/good.zip). These two captures are for the same client (same client-id), the second one was triggered by application backoff logic which start a new client with exponential backoff and eventually succeeded.

I attach here also paho  logs for one failed session 
[paho0.fail.log](https://github.com/eclipse/paho.mqtt.java/files/4482547/paho0.fail.log)


One of the clients failed 28 times over 15 minutes until application finally succeeding. 

typical stacktrace of the failed connection is:

 
2020-04-15T14:37:03.067441372Z Connection lost (32109) - java.io.IOException: Connection is lost.
2020-04-15T14:37:03.067491226Z  at org.eclipse.paho.client.mqttv3.internal.CommsReceiver.run(CommsReceiver.java:197)
2020-04-15T14:37:03.067500361Z  at java.lang.Thread.run(Thread.java:748)
2020-04-15T14:37:03.070894261Z Caused by: java.io.IOException: Connection is lost.
2020-04-15T14:37:03.070929348Z  at org.eclipse.paho.client.mqttv3.internal.CommsReceiver.run(CommsReceiver.java:172)

if I comment out the fix for issue 719 (https://github.com/eclipse/paho.mqtt.java/blob/8b572c635b63661b7cc371a4e0d8a536e831fcaa/org.eclipse.paho.client.mqttv3/src/main/java/org/eclipse/paho/client/mqttv3/internal/CommsReceiver.java#L170) then the clients connect without any failure

**Comment by marcostorto** [comment]
Unfortunately I was not able to reproduce the issue in a simpler setup, I suspect that the issue is triggered by CPU overload as server CPU get close to 100% usage for several minutes after applicaitn restart in this scenario, as both consumer application and broker run on the same server

**Comment by marcostorto** [comment]
My understanding of the issue is:

1. paho creates TCP socket with 1000ms socket timeout (TCPNetworkModule.java)
2. if socket timeout expires in readMqttWireMessage() then SocketTimeoutException is catched and return null message
3. if socket timeout happens during initial connection (e.g. due to CPU overload) then the fix  for issue 719 causes connection termination, while without this fix connection completes normally.

Basically the current fix for issue 719 re-use socket read timeout for the additional purpose of detecting a hung MQTT handshake, but if any other cause triggers a socket read timeout during connection this fix causes unneded disconnections. 
A better solution would be to remove current fix and add a separate timeout to wait for reception of MQTT connection ack.

**Comment by rdasgupt** [comment]
@marcostorto 
Thanks for providing detailed information from your test case. I will go thru the provided logs.

Going thru your details, there is a possibility that your test server is overloaded and may not return acks before socket timeout in clients. I think your problem will be resolved if you increase connection timeout from 1000 ms (i.e. 1 second) to 60000 ms (60 seconds). 

In our scale test environment, we have 100 paho client instances per JVM. And we load our server with about 1 million clients. Our server is different though. We use IBM MessageGateway broker in our scale test environment. We don't see any problem. 

Anyway, please try with latest released client 1.2.3 and increase timeout from 1000 to 60000.
Meanwhile, let me check again if we can rework fix for 719.

**Comment by marcostorto** [comment]
The issue is not linked with connection timeout, TCP connection completes normally and paho sends MQTT connect packet.
Issue is due to socket timeout, which is hard wired in paho code:
https://github.com/eclipse/paho.mqtt.java/blob/55137d947418c4e0e74326eb9d080e5f1fd9a473/org.eclipse.paho.client.mqttv3/src/main/java/org/eclipse/paho/client/mqttv3/internal/TCPNetworkModule.java#L75

Should I recompile a modified paho with longer timeout?

**Comment by rdasgupt** [comment]
I think increasing timeout in setSoTimeout() may not be a good idea. But you can try if it helps in anyways. I am checking if fix for 719 can be improved.

**Comment by rdasgupt** [comment]
@marcostorto 
So far I couldn't find anything wrong with fix for 719. We have tried with connection storm tests with large number of clients publishing messages and eventually all clients got connected successfully. The CPU usage on the IBM MessageGateway server 100%. 

I don't know how verneMQ handles client connections, so can't comment on connection storm case or client connection in general on verneMQ, when server is over loaded. In MessageGateway, we have dedicated threads to process client connections so it scales well.

I will revisit this again early next week, meanwhile I would suggest you to check with verneMQ if there is any configuration that can help in your use case.

**Comment by ioolkos** [comment]
@marcostorto try setting `max_inflight_messages = 200` and see what this does for you.
@rdasgupt this is not about a connect storm, I think. It's about delivering lots of on-disk messages to a few clients after they have reconnected.

So you only test this open source client against a closed source broker? ;) You should check out VerneMQ, it's awesome :)

**Comment by rdasgupt** [comment]
@ioolkos 
Thanks for following this issue and providing your comments.

Though the use case @marcostorto is trying is not directly related to connection storm, but you can have the same problem with connection storm if server is overloaded due to processing connections. If server doesn't respond to initial connection before socket times out, the client will be in connect-disconnect loop. But eventually server should process all connections. In use case that @marcostorto is trying is about connection getting in connect-disconnect loop when server is too busy to process initial connection. 

I haven't worked with VerneMQ yet. Probably I will spend some time to check that out (when I get some time).

**Comment by marcostorto** [comment]
I'll try max_inflight_messages = 200, I guess it's going to help but unfortunately I cannot use this setting as it's a global parameter in vernemq and some other clients in the system need a larger value of max_inflight_messages to cope with high-throughput traffic.

Anyway I think the fix for 719 is not *wrong*, but it sets a fixed timeout of 1 second for connection handshake which is not enough in my application. What I think is needed is to have a configurable timeout for connection handshake, so that it can be set by library users.

I'll try to prepare a PR for this, but I cannot commit on a hard timeline.

**Comment by marcostorto** [comment]
I was finally able to re-test with 1.2.5 and I see the issue is not present anymore, due to a small change in https://github.com/eclipse/paho.mqtt.java/blob/6f35dcb785597a6fd49091efe2dba47513939420/org.eclipse.paho.client.mqttv3/src/main/java/org/eclipse/paho/client/mqttv3/internal/CommsReceiver.java#L171
